### PR TITLE
gcp - added spanner-instance-backup

### PR DESCRIPTION
**Pull request by evhenyshaliov**
Use case:

`
policies:
  - name: spanner_backup_expiration_time_30_days_or_more
    description: |
      Cloud Spanner backup is created with an expiration date of 29 days or less
    resource: gcp.spanner-instance-backup
    filters:
      - type: time-range
        value: 29
`

### CODE REVIEW

File: tools/c7n_gcp/tests/test_query.py [hunk 1]
```diff
@@ -35,6 +35,7 @@ def test_gcp_resource_metadata_asset_type():
         'ml-model',
         'notebook',
         'sourcerepo',
+        'spanner-instance-backup',
```
Comment (kapilt) [inline tools/c7n_gcp/tests/test_query.py:38]: please don't add to this list, if the resource is supported by asset inventory, just add the metadata to the resource type

[spanner.googleapis.com/Instance](https://cloud.google.com/spanner/docs/reference/rest/v1/projects.instances)

File: tools/c7n_gcp/c7n_gcp/resources/spanner.py [hunk 2]
```diff
@@ -44,6 +45,37 @@ def get_label_params(resource, all_labels):
                         'field_mask': ', '.join(['labels'])}}
 
 
+@resources.register('spanner-instance-backup')
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/spanner.py:48]: s/spanner-instance-backup/spanner-backup

File: tools/c7n_gcp/tests/test_query.py (hunk 1 above)
Comment (evhenyshaliov) [reply tools/c7n_gcp/tests/test_query.py]: Fixed. Please check

File: tools/c7n_gcp/tests/test_spanner.py [hunk 3]
```diff
@@ -357,3 +357,21 @@ def test_set_iam_policy_remove_bindings_star(self):
...
+
+class TestIamSpannerInstance(BaseTest):
+
+    def test_query(self):
+        project_id = 'cloud-custodian'
+        session_factory = self.replay_flight_data('test-iam-spanner-instance',
+                                                  project_id=project_id)
+        policy = self.load_policy(
+            {'name': 'iam-spanner-instance',
+             'resource': 'gcp.spanner-instance'},
```
Comment (kapilt) [inline tools/c7n_gcp/tests/test_spanner.py:370]: this isn't testing query against a backup, but a spanner instance

Comment (evhenyshaliov) [reply tools/c7n_gcp/tests/test_spanner.py]: right, sorry. It's fixed

File: tools/c7n_gcp/c7n_gcp/resources/spanner.py (hunk 2 above)
Comment (evhenyshaliov) [reply tools/c7n_gcp/c7n_gcp/resources/spanner.py:49]: could you provide more details? should we add aliases?

Comment (kapilt) [reply tools/c7n_gcp/c7n_gcp/resources/spanner.py:49]: I think just change the resource name to `spanner-backup` which matches what gcp does for it in asset inventory.

File: tools/c7n_gcp/c7n_gcp/resources/spanner.py [hunk 4]
```diff
@@ -44,6 +46,87 @@ def get_label_params(resource, all_labels):
...
+
+
+@SpannerInstanceBackup.filter_registry.register('time-range')
+class SpannerInstanceBackupTimeRangeFilter(Filter):
+    """Filters bigtable instance clusters backups based on a time range
+
+    .. code-block:: yaml
+
+        policies:
+          - name: bigtable_backup_expiration_time_30_days
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/spanner.py:85]: wrong policy name

File: tools/c7n_gcp/c7n_gcp/resources/spanner.py [hunk 5]
```diff
@@ -44,6 +46,87 @@ def get_label_params(resource, all_labels):
...
+
+    create_time_field_name = 'createTime'
+    expire_time_field_name = 'expireTime'
+
+    def process(self, resources, event=None):
+        filtered_resources = []
+        value = self.data.get('value')
+
+        for resource in resources:
+            create_time_pattern = self.datetime1_pattern
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/spanner.py:108]: why is this logic being duplicated?

File: tools/c7n_gcp/c7n_gcp/resources/spanner.py [hunk 6]
```diff
@@ -44,6 +46,87 @@ def get_label_params(resource, all_labels):
...
+            description: |
+              Cloud Bigtable backup expiration time is 29 days or less
+            resource: gcp.bigtable-instance-cluster-backup
+            filters:
+              - type: time-range
+                value: 29
+    """
+    schema = type_schema('time-range',
+                         value={'$ref': '#/definitions/filters_common/value'})
+    permissions = ('bigtable.backups.list',)
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/spanner.py:95]: wrong permissions

File: tools/c7n_gcp/c7n_gcp/resources/spanner.py (hunk 2 above)
Comment (evhenyshaliov) [reply tools/c7n_gcp/c7n_gcp/resources/spanner.py]: fixed

File: tools/c7n_gcp/c7n_gcp/resources/spanner.py (hunk 4 above)
Comment (evhenyshaliov) [reply tools/c7n_gcp/c7n_gcp/resources/spanner.py]: sorry, my bad. It's fixed

File: tools/c7n_gcp/c7n_gcp/resources/spanner.py (hunk 6 above)
Comment (evhenyshaliov) [reply tools/c7n_gcp/c7n_gcp/resources/spanner.py]: fixed

File: tools/c7n_gcp/c7n_gcp/resources/spanner.py (hunk 5 above)
Comment (evhenyshaliov) [reply tools/c7n_gcp/c7n_gcp/resources/spanner.py]: added a new common filter and re-used it from bigtable and spanner

**Review by kapilt — APPROVED** [review]
thanks for the changes, lgtm, its currently failing ci though on the doc build (warnings treated as errors), this could be a ci doc build cache consequence, unclear atm. I'll investigate a bit more, the warnings in question fwiw,

```
/home/runner/work/cloud-custodian/cloud-custodian/docs/source/gcp/resources/group-bigtableadmin.rst: WARNING: document isn't included in any toctree
checking consistency... done
/home/runner/work/cloud-custodian/cloud-custodian/docs/source/gcp/resources/group-spanner.rst: WARNING: document isn't included in any toctree
```

**Comment by kapilt** [comment]
hmmm, now this is getting import errors on the spanner module, which seems to be the underlying issue vs the doc build.

**Comment by kapilt** [comment]
I can trigger a similar error locally, which does seem to be due to an import cycle. the movement of time range to filters seemed to cause so likely its the source on the imports.

``` 
Traceback (most recent call last):
  File "/Users/kapilt/projects/src-custodian/tools/c7n_gcp/c7n_gcp/resources/spanner.py", line 3, in <module>
    from c7n.utils import type_schema, local_session
  File "/Users/kapilt/projects/src-custodian/c7n/utils.py", line 10, in <module>
    import logging
  File "/Users/kapilt/projects/src-custodian/tools/c7n_gcp/c7n_gcp/resources/logging.py", line 3, in <module>
    from c7n.utils import local_session, type_schema
ImportError: cannot import name 'local_session' from partially initialized module 'c7n.utils' (most likely due to a circular import) (/Users/kapilt/projects/src-custodian/c7n/utils.py)
```

**Comment by evhenyshaliov** [comment]
it's fixed. Sorry for delay.

**Review by kapilt — APPROVED** [review]
lgtm

**Pull request merged**